ddtrace/internal/forksafe.py

```diff
@@ -0,0 +1,106 @@
+import inspect
+import functools
+import os
+import threading
+
+
+__all__ = [
+    "call_nocheck",
+    "ddtrace_after_in_child",
+    "register",
+]
+
+
+BUILTIN_FORK_HOOKS = hasattr(os, "register_at_fork")
+
+if BUILTIN_FORK_HOOKS:
+    registry = []
+
+    def ddtrace_after_in_child():
+        global registry
+
+        for fn, hook in registry:
+            hook()
+            fn._fork_occurred = True
+
+    def register(after_in_child):
+        # TODO: this function won't work if not used as a decorator.
+        # eg: register(hook) won't result in hook being called.
+        def wrapper(func):
+            args = inspect.getfullargspec(func)[0]
+            include_arg = "is_in_child_after_fork" in args
+
+            @functools.wraps(func)
+            def fsfunc(*args, **kwargs):
+                fork_occurred = fsfunc._fork_occurred
+                fsfunc._fork_occurred = False
+                if include_arg:
+                    return func(*args, is_in_child_after_fork=fork_occurred, **kwargs)
+                else:
+                    return func(*args, **kwargs)
+
+            fsfunc._fork_occurred = False
+            registry.append((fsfunc, after_in_child))
+            return fsfunc
+
+        return wrapper
+
+    def call_nocheck(f, *args, **kwargs):
+        return f(*args, **kwargs)
+
+    os.register_at_fork(after_in_child=ddtrace_after_in_child)
+else:
+    PID = os.getpid()
+    PID_LOCK = threading.Lock()
+    registry = []
+
+    def ddtrace_after_in_child():
+        global registry
+
+        for hook in registry:
+            hook()
```

Comment:
What happens in case a hook fails?

Reply:
I think it's fine to let the exception bubble up. We could ensure that all hooks are called by catching and re-raising but I think it should be the responsibility of the user that the hook function will not raise.

If we did permit raising then it becomes quite strange to catch those errors. Same as for the built-in hooks... is it even possible to handle exceptions raised in that case? I think I've seen the Python stdlib handle this kind of thing by catching and logging a WARNING-level message with the stack trace.

What do you think? I don't have any strong opinions on this.

Reply:
I think we should mimic what the built-in hook does.